lib/cn/cn_tree_cursor.c

```diff
@@ -0,0 +1,774 @@
+ * two layers of bin heaps.
+ *
+ * Each node (root node and the one leaf node) uses a binheap to merge keys across all its kvsets.
+ * These two binheaps feed into a top level binheap that forms the cn cursor.
+ *
+ * Each node uses a struct cn_level_cursor to iterate over its kvsets.
+ */
+
+#include <hse_util/event_counter.h>
+#include <hse_util/page.h>
+#include <hse_util/logging.h>
+#include <hse_util/assert.h>
+#include <hse_util/table.h>
+#include <hse_util/key_util.h>
+#include <hse_util/keycmp.h>
+#include <hse_util/bin_heap.h>
+
+#include <hse_ikvdb/cn.h>
+
+#include <cn/cn_cursor.h>
+
+#include "route.h"
+
+#define MTF_MOCK_IMPL_cn_tree_cursor
+
+#include "cn_tree_internal.h"
+#include "cn_tree_cursor.h"
+#include "kvset_internal.h"
+#include "kvset.h"
+#include "kv_iterator.h"
+
+/*
+ * Min heap comparator for forward iteration.
+ *
+ * Returns:
+ *   < 0 : a_blob < b_blob
+ *   > 0 : a_blob > b_blob
+ *  == 0 : a_blob == b_blob
+ */
+static int
+cn_kv_cmp(const void *a_blob, const void *b_blob)
+{
+    const struct cn_kv_item *a = a_blob;
+    const struct cn_kv_item *b = b_blob;
+
+    return key_obj_cmp(&a->kobj, &b->kobj);
+}
+
+/*
+ * Max heap comparator for reverse iteration.
+ * with a caveat: A ptomb sorts before all keys w/ matching
+ * prefix.
+ *
+ * Returns:
+ *   < 0 : a_blob > b_blob
+ *   > 0 : a_blob < b_blob
+ *  == 0 : a_blob == b_blob
+ */
+static int
+cn_kv_cmp_rev(const void *a_blob, const void *b_blob)
+{
+    const struct cn_kv_item *a = a_blob;
+    const struct cn_kv_item *b = b_blob;
+    size_t                   a_klen = a->kobj.ko_pfx_len + a->kobj.ko_sfx_len;
+    size_t                   b_klen = b->kobj.ko_pfx_len + b->kobj.ko_sfx_len;
+
+    int rc;
+
+    if (!(a->vctx.is_ptomb ^ b->vctx.is_ptomb))
+        return key_obj_cmp(&b->kobj, &a->kobj);
+
+    /* Exactly one of a and b is a ptomb. */
+    if (a->vctx.is_ptomb && a_klen <= b_klen) {
+        rc = key_obj_ncmp(&b->kobj, &a->kobj, a_klen);
+        if (rc == 0)
+            return -1; /* a wins */
+    } else if (b->vctx.is_ptomb && b_klen <= a_klen) {
+        rc = key_obj_ncmp(&b->kobj, &a->kobj, b_klen);
+        if (rc == 0)
+            return 1; /* b wins */
+    }
+
+    /* Non-ptomb key is shorter than ptomb. Full key compare. */
+    return key_obj_cmp(&b->kobj, &a->kobj);
+}
+
+MTF_STATIC merr_t
+cn_tree_kvset_refs(struct cn_tree_node *node, struct cn_level_cursor *lcur)
+{
+    struct table *tab = lcur->cnlc_kvref_tab;
+    struct kvset_list_entry *le;
+
+    lcur->cnlc_dgen_hi = lcur->cnlc_dgen_lo = 0;
+    table_reset(tab);
+
+    list_for_each_entry (le, &node->tn_kvset_list, le_link) {
+        struct kvset *kvset = le->le_kvset;
+        struct kvref *k;
+        u64 dgen = kvset_get_dgen(kvset);
+
+        if (!lcur->cnlc_dgen_hi)
+            lcur->cnlc_dgen_hi = dgen;
+
+        k = table_append(tab);
+        if (ev(!k))
+            return merr(ENOMEM);
```

Comment:
Is it legit to return with some refs but not all refs?

Reply:
No, this error path needs to call `cn_lcur_kvset_release()`. Fixed it. But in other places, the caller would check for the error and call `cn_tree_cursor_destroy()`, which would release all the resources.